Feature/profile search impr

ilililililililililililililil opened the pull request:
If apply these commits will:
- Add several fixes and improvements for EP /api/v1/profiles/ profiles 
- Add tests for EP /api/v1/profiles/ profiles search

Commit:
Add several fixes and improvements for EP /api/v1/profiles/ profiles search
by location, sex, age.

Commit:
Add tests for EP /api/v1/profiles/ profiles search

ilililililililililililililil committed:
Merge branch 'develop' into feature/profile-search-impr

Commit:
Add filter fix for EP /api/v1/profiles/

Commit:
Add tests fix for EP /api/v1/profiles
add import style fix

Commit:
Add several fixes and improvements for EP /api/v1/profiles/
- add viewers exclude
- add requester profile exclude

Commit:
Add several fixes and improvements for EP /api/v1/profiles/
- add filters location by name field fix

Commit:
Add tests fix for EP /api/v1/profiles
- add 4 search criteria
- add wrong methods check tests

ilililililililililililililil commented:
@TomatoInOil посмотри пожалуйста PR

TomatoInOil commented on src/backend/search/tests/test_views.py:
```diff
@@ -183,3 +183,144 @@ def test_matches_correct_data(self):
+        cls.test_user_9 = UserFromTelegram.objects.create(telegram_id=9)
+        cls.test_user_10 = UserFromTelegram.objects.create(telegram_id=10)
+        cls.sex_m = "Парень"
+        cls.sex_f = "Девушка"
```
Будет хорошо использовать этот класс:
https://github.com/Studio-Yandex-Practicum/coliving_bot/blob/59334ad503d35a98e80cd7ab299fac8184ce4ceb/src/backend/profiles/constants.py#L13-L19
Заменив на `Sex.MAN`, `Sex.WOMAN`, там где `cls.sex_m`, `cls.sex_f`, `self.sex_m`, `self.sex_f`.

TomatoInOil commented on src/backend/search/tests/test_views.py:
```diff
@@ -183,3 +183,144 @@ def test_matches_correct_data(self):
+        cls.test_user_7 = UserFromTelegram.objects.create(telegram_id=7)
+        cls.test_user_8 = UserFromTelegram.objects.create(telegram_id=8)
+        cls.test_user_9 = UserFromTelegram.objects.create(telegram_id=9)
+        cls.test_user_10 = UserFromTelegram.objects.create(telegram_id=10)
```
Здесь можно использовать [list comprehension](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions) и далее получать пользователей по индексу.
```Python
test_users = [UserFromTelegram.objects.create(telegram_id=id) for id in range(1, 11)]

# Использование
test_users[0]
```

TomatoInOil commented on src/backend/search/tests/test_views.py:
```diff
@@ -183,3 +183,144 @@ def test_matches_correct_data(self):
+        cls.profile_10 = Profile.objects.create(
+            user=cls.test_user_10, name="Name_10", age=44, location=cls.location_s,
+            sex=cls.sex_f
+        )
```
Можно определить кортеж тестовых данных со значениями полей, перебрав их, создать профили, используя тот же list comprehension. 
```Python
ProfileTestData = namedtuple("ProfileTestData", ["user", "name", "age", "location", "sex"])

# ...

profiles_data = (
    ProfileTestData(test_users[0], "Name_1", 21, cls.location_m, Sex.MAN),
    ProfileTestData(test_users[1], "Name_2", 25, cls.location_m, Sex.MAN),
    # И тд
)
test_profiles = [
    Profile.objects.create(**data._asdict()) for data in profiles_data
]
```

TomatoInOil commented on src/backend/search/tests/test_views.py:
```diff
@@ -183,3 +183,144 @@ def test_matches_correct_data(self):
+             {"user": 5, "name": "Name_5", "sex": cls.sex_f, "age": 34,
+              "location": "M", "about": None, "is_visible": True, "images": []},
+             ]
+        }
```
Здесь дублируется то, что мы передаём при создании профиля, но для отдельных пользователей. Так почему бы нам не использовать сами профили, а не их вводные данные? Нам же нужно проверить корректность фильтрации, а не сериализации полей. 
```Python
cls.expected_search_data_2 = {
    "search_criteria": {
        "location": cls.LOCATION_M_NAME,
        "sex": Sex.WOMAN,
        "age_min": "30",
        "age_max": "40",
    },
    "search_result": [
        ProfileSerializer(test_profiles[3]).data,
        ProfileSerializer(test_profiles[4]).data,
    ],
}
```

TomatoInOil commented on src/backend/search/tests/test_views.py:
```diff
@@ -183,3 +183,144 @@ def test_matches_correct_data(self):
+
+class ProfileSearchViewTests(APITestCase):
+    """Тесты для ProfilesSearchView."""
+
+    empty_match_data = []
```
Далее идёт повторное определение в `def setUpTestData(cls)`, нужно оставить что-то одно. При наименовании констант используется upper-case https://stackoverflow.com/a/45702782
Но есть другая проблема, эта константа не используется далее в тестах.

ilililililililililililililil replied:
это не было константой, лишний текст

TomatoInOil commented on src/backend/search/tests/test_views.py:
```diff
@@ -183,3 +183,144 @@ def test_matches_correct_data(self):
+        }
+        cls.expected_search_data_4 = {
+            "search_criteria":
+            {"location": "S", "sex": cls.sex_f, "age_min": "40", "age_max": "50"},
```
Хорошо будет вынести названия городов в константы класса.

TomatoInOil commented on src/backend/search/views.py:
```diff
@@ -46,25 +46,24 @@ def get_queryset(self):
+        req_user = Profile.objects.get(pk=user.pk)
+        excl_list.append(req_user.pk)
+        for viwed in Profile.objects.all().filter(viewers=user):
+            excl_list.append(viwed.pk)
```
1. Вместо того, чтобы совершать ещё один запрос к БД `Profile.objects.get(pk=user.pk)`, можно использовать в фильтрации условие, где поле `user` в профиле соответствует `user`, полученному из параметра `viewer`
```Python
Q(user=user)
```
2. Используя оператор [| (OR)](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#or) можно объединить эти два `queryset` при фильтрации в один
```Python
Profile.objects.filter(
    Q(user=user) | Q(viewers=user)
)
```
3. Вместо цикла для формирования списка с pk можно использовать [.values_list()](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#values-list)

Применив всё это, получим:
```Python
excl_list = Profile.objects.filter(
    Q(user=user) | Q(viewers=user)
).values_list("pk", flat=True)
```

TomatoInOil reviewed (changes requested):
В основном рекомендации направлены на уменьшение ненужного дублирования в коде и улучшение его читаемости, поддерживаемости. 

Если говорить про тесты, то в `global_search_results_data` есть случай, в котором тесты проверяют, что профиль `viewer` не попадает в ответ, но лично мне захотелось бы добавить тест, который проверят это более явно, исключив влияние остальных фильтров. 
Например, передав в kwargs только viewer 
`response = self.client.get(self.url_reverse, dict(viewer=viewer_telegram_id))`
И с помощью `assertNotIn` проверить, что переданный `viewer_telegram_id` не попадает в `[profile["user"] for profile in response.json()]`

Commit:
Add tests, view fix for EP /api/v1/profiles
- add 4 search criteria
- add wrong methods check tests
- correct review issues
